drivers/HighResClock.h

```diff
@@ -0,0 +1,113 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef MBED_HIGHRESCLOCK_H
+#define MBED_HIGHRESCLOCK_H
+
+#include <chrono>
+#include <mstd_mutex>
+#include "hal/us_ticker_api.h"
+#include "platform/mbed_power_mgmt.h"
+
+namespace mbed {
+/**
+ * \defgroup drivers_HighResClock HighResClock class
+ * \ingroup drivers-public-api-ticker
+ * @{
+ */
+
+/**
+ * A C++11 Clock representing the HAL us_ticker.
+ * This version does not lock deep sleep to keep the clock running - the
+ * clock will pause whenever deep sleep is entered.
+ *
+ * Lock and unlock methods are provided to support code that wants to control the
+ * deep sleep to keep the clock running. LowPowerClock provides the same
+ * methods as dummies to support generic code that works with either clock.
+ *
+ * If code is not in a position to use the lock/unlock methods, use
+ * SimpleHighResClock instead.
```

Comment:
Yep, that was previous naming attempt. `Simple` being meant to imply it had some big drawback! "Simple" things always have a downside...

Reply:
"Limited"?

Reply:
Well, it's not limited. On the contrary, it's unlimited! Just keeps running...

Reply:
Unsafe? In Swift it means that you have to manage everything yourself and take care of cleaning up when you leave.

Reply:
It's not really that either, in that it wouldn't be possible to clean up in this form. There's no tracking of how many people are using it, or the option to say you're not any more - that would be the `HighResClock`.

Reply:
Hmm :D then I tend to agree with @bulislaw when he says

> Is that really needed? I'm just afraid it it pushes people towards wrong use patterns.

Reply:
Yeah, I invented it while writing the design document. It's the answer to the question "what's the direct equivalent of `std::high_resolution_clock`?

I originally had it as an example in there, then decided I may as well put it in the code, rather than the Markdown.

I can push it back to just being in the Markdown.